File: src/AI/Rystem.PlayFramework.Client/src/rystem/src/models/PlayFrameworkRequest.ts
import { ContentItem } from "./ContentItem";
import { ClientInteractionResult } from "./ClientInteractionResult";

/**
 * Execution mode for the scene.
 */
export type SceneExecutionMode = "Direct" | "Planning" | "DynamicChaining" | "Scene";

/**
 * Cache behavior for the request.
 */
export type CacheBehavior = "Default" | "Avoidable" | "Forever";

/**
 * Request-specific settings matching C# SceneRequestSettings.
 */
export interface SceneRequestSettings {
    /** Maximum recursion depth for planning (default: 5). */
    maxRecursionDepth?: number;
    /** Execution mode: Direct, Planning, DynamicChaining, Scene. */
    executionMode?: SceneExecutionMode;
    /** Whether to enable summarization. */
    enableSummarization?: boolean;
    /** Whether to enable director for multi-scene orchestration. */
    enableDirector?: boolean;
    /** Model to use for this request (overrides default). */
    modelId?: string;
    /** Temperature for LLM calls (0.0 - 2.0). */
    temperature?: number;
    /** Maximum tokens to generate. */
    maxTokens?: number;
    /** Cache behavior for this request. */
    cacheBehavior?: CacheBehavior;
    /** Maximum budget for this request. Null for unlimited. */
    maxBudget?: number;
    /** Enable streaming for text responses. */
    enableStreaming?: boolean;
    /** Maximum number of scenes in dynamic chaining mode (default: 5). */
    maxDynamicScenes?: number;
    /** Unique key for this conversation (used by cache and memory). */
    conversationKey?: string;
    /** Name of the scene to execute directly (used with SceneExecutionMode "Scene"). */
    sceneName?: string;
    /** Results from client-side tool executions. */
    clientInteractionResults?: ClientInteractionResult[];
}

/**
 * Request model for PlayFramework HTTP API.
 * Matches C# PlayFrameworkRequest contract.
 */
export interface PlayFrameworkRequest {
    /**
     * User message (text).
     */
    message?: string;

    /**
     * Multi-modal content items (images, audio, video, files, URIs).
     */
    contents?: ContentItem[];

    /**
     * Custom metadata (userId, sessionId, etc.).
     */
    metadata?: Record<string, any>;

    /**
     * Request-specific settings (override defaults).
     */
    settings?: SceneRequestSettings;

    /**
     * Conversation key for resuming execution after client interaction.
     * When present, server loads conversation state from cache.
     */
    conversationKey?: string;

    /**
     * Results from client-side tool executions.
     * Used to resume execution after AwaitingClient status.
     */
    clientInteractionResults?: ClientInteractionResult[];
}

File: src/AI/Rystem.PlayFramework/Api/Models/PlayFrameworkRequest.cs
using Microsoft.Extensions.AI;

namespace Rystem.PlayFramework.Api.Models;

/// <summary>
/// Request model for PlayFramework HTTP API.
/// </summary>
public sealed class PlayFrameworkRequest
{
    /// <summary>
    /// User message (text).
    /// </summary>
    public string? Message { get; set; }

    /// <summary>
    /// Multi-modal contents (images, audio, files).
    /// Use this for advanced scenarios with multiple content types.
    /// </summary>
    public List<ContentItem>? Contents { get; set; }

    /// <summary>
    /// Request metadata (userId, tenantId, sessionId, etc.).
    /// Used for rate limiting, telemetry, and custom logic.
    /// </summary>
    public Dictionary<string, object>? Metadata { get; set; }

    /// <summary>
    /// Request-specific settings (override defaults).
    /// </summary>
    public SceneRequestSettings? Settings { get; set; }

    /// <summary>
    /// Conversation key for resuming execution after client interaction.
    /// When present with ClientInteractionResults, server loads continuation state from cache.
    /// </summary>
    public string? ConversationKey { get; set; }

    /// <summary>
    /// Results from client-side tool executions.
    /// Used to resume execution after AwaitingClient status.
    /// </summary>
    public List<ClientInteractionResult>? ClientInteractionResults { get; set; }
}

/// <summary>
/// Represents a single content item (text, image, audio, file, URI).
/// </summary>
public sealed class ContentItem
{
    /// <summary>
    /// Content type: "text", "image", "audio", "video", "file", "uri"
    /// </summary>
    public string Type { get; set; } = "text";

    /// <summary>
    /// Text content (for type "text").
    /// </summary>
    public string? Text { get; set; }

    /// <summary>
    /// Base64-encoded content (for type "image", "audio", "video", "file").
    /// </summary>
    public string? Data { get; set; }

    /// <summary>
    /// Media type / MIME type (e.g., "image/png", "audio/mp3").
    /// </summary>
    public string? MediaType { get; set; }

    /// <summary>
    /// URI content (for type "uri").
    /// </summary>
    public string? Uri { get; set; }

    /// <summary>
    /// Optional name/filename.
    /// </summary>
    public string? Name { get; set; }

    /// <summary>
    /// Converts to AIContent for PlayFramework execution.
    /// </summary>
    internal AIContent ToAIContent()
    {
        return Type.ToLowerInvariant() switch
        {
            "text" => new TextContent(Text ?? string.Empty),
            "image" or "audio" or "video" or "file" => new DataContent(
                Convert.FromBase64String(Data ?? string.Empty),
                MediaType),
            "uri" => new UriContent(new Uri(Uri ?? "about:blank"), MediaType),
            _ => throw new InvalidOperationException($"Unsupported content type: {Type}")
        };
    }
}
